src/PureSMT/Process.hs

```diff
@@ -5,119 +5,62 @@
-    debugMode :: Bool,
-    queue :: IORef Builder -- only used in non-debug mode
+  { backend :: Bck.Solver,
+    debugMode :: Bool
   }
 
--- | Create a brand-new context for Z3 to work in.
-initZ3Instance ::
+-- | Launch a solver.
+-- Here we just initialize a new context for the Z3 C API to work with.
+launchSolver ::
   -- | Whether or not to debug the interaction
   Bool ->
   IO Solver
-initZ3Instance dbg = do
-  let ctxFinalizer =
-        [C.funPtr| void free_context(Z3_context ctx) {
-                                      Z3_del_context(ctx);
-                                      } |]
-  solverCtx <-
-    newForeignPtr ctxFinalizer
-      =<< [CU.block| Z3_context {
-                     Z3_config cfg = Z3_mk_config();
-                     Z3_context ctx = Z3_mk_context(cfg);
-                     Z3_del_config(cfg);
-                     return ctx;
-                     } |]
-  solverQueue <- newIORef mempty
-  let solver = Solver solverCtx dbg solverQueue
-  when (debugMode solver) $ do
-    -- this should not be enabled in non-debug mode, as it messes with parsing
-    -- the outputs of commands that are actually interesting
-    setOption solver ":print-success" "true"
-  setOption solver ":produce-models" "true"
-  return solver
-
--- | Send a bytestring to Z3.
--- This function is thread-safe as long as concurrent instances do not share the
--- same logical context.
-send :: Solver -> BS.ByteString -> IO SExpr
-send solver cmd = do
-  let ctx = context solver
-  when (debugMode solver) $ do
-    BS.putStrLn $ "[send] " `BS.append` cmd
-  resp <-
-    [CU.exp| const char* {
-         Z3_eval_smtlib2_string($fptr-ptr:(Z3_context ctx), $bs-ptr:cmd)
-         } |]
-      >>= BS.packCString
-  case readSExpr resp of
-    Nothing -> do
-      fail $ "solver replied with:\n" ++ BS.unpack resp
-    Just (sexpr, _) -> do
-      when (debugMode solver && sexpr /= Atom "success") $ do
-        putStrLn $ "[recv] " ++ showsSExpr sexpr ""
-      return sexpr
-
--- | Push a command on the solver's queue of commands to evaluate.
--- The command must not produce any output when evaluated.
-putQueue :: Solver -> SExpr -> IO ()
-putQueue solver expr = do
-  let solverQueue = queue solver
-  cmds <- readIORef solverQueue
-  writeIORef solverQueue $ cmds <> renderSExpr expr
-
--- | Empty the queue of commands to evaluate and return its content as a bytestring.
-flushQueue :: Solver -> IO (BS.ByteString)
-flushQueue solver = do
-  let solverQueue = queue solver
-  cmds <- readIORef solverQueue
-  writeIORef solverQueue mempty
-  return $ serializeBatch cmds
-
--- | Have Z3 evaluate a command in SExpr format.
+launchSolver dbg = do
+  z3 <- Z3.toBackend <$> Z3.new
+  solver <- Bck.initSolver z3 lazy
+  return $ Solver solver dbg
+  where
+    lazy = dbg
+
+-- | Have the evaluate a command in SExpr format.
 -- This forces the queued commands to be evaluated as well, but their results are
 -- *not* checked for correctness.
```

Comment:
If I'm not mistaken, there are no queued commands anymore? Or at least not visible from PureSMT's point of view?

Reply:
They are still visible in that we manually enable queuing inside the `launchSolver` function, but we could decide to get rid of this comment yes.

Reply:
Is that the `lazy` parameter of `Bck.initSolver`? Otherwise, when reading the code of `launchSolver`, I don't really see that there is a queue somewhere.

Reply:
It was, I made its name more explicit in https://github.com/tweag/pirouette/pull/162/commits/80a850a02cc1953582ac616bbc16e0e21caa7d4a.

Reply:
As well as fixing a bug, I see? Great!